rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/rascal/RascalLanguageServices.java

```diff
@@ -235,6 +245,115 @@ public InterruptibleFuture<ITuple> getRename(ITree module, Position cursor, Set<
     }
 
+    private ISourceLocation sourceLocationFromUri(String uri) {
+        try {
+            return URIUtil.createFromURI(uri);
+        } catch (URISyntaxException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private Optional<ISourceLocation> findContainingWorkspaceFolder(ISourceLocation loc, List<ISourceLocation> workspaceFolders) {
+        return workspaceFolders.stream()
+            .filter(folderLoc -> URIUtil.isParentOf(folderLoc, loc))
+            .findFirst();
+    }
+
+    private ISet qualfiedNameChangesFromRenames(List<FileRename> renames, Set<ISourceLocation> workspaceFolders, Function<ISourceLocation, PathConfig> getPathConfig) {
+        // Sort workspace folders so we get the most specific folders first
+        List<ISourceLocation> sortedWorkspaceFolders = workspaceFolders.stream()
+            .sorted((o1, o2) -> o1.toString().compareTo(o2.toString()))
+            .collect(Collectors.toList());
+
+        var writer = VF.setWriter();
+        for (var rename : renames) {
+            ISourceLocation currentLoc = sourceLocationFromUri(rename.getOldUri());
+            ISourceLocation newLoc = sourceLocationFromUri(rename.getNewUri());
+
+            ISourceLocation currentWsFolder = findContainingWorkspaceFolder(currentLoc, sortedWorkspaceFolders)
+                .orElseThrow(() -> new RuntimeException(String.format("Cannot automatically change uses of %s, since it is outside the current workspace.", currentLoc)));
+
+            ISourceLocation newWsFolder = findContainingWorkspaceFolder(newLoc, sortedWorkspaceFolders)
+                .orElseThrow(() -> new RuntimeException(String.format("Cannot automatically change uses of %s, since that it is outside the current workspace.", newLoc)));
+
+            if (!currentWsFolder.equals(newWsFolder)) {
+                String commonProjPrefix = StringUtils.getCommonPrefix(currentWsFolder.toString(), newWsFolder.toString());
+                String currentProject = StringUtils.removeStart(currentWsFolder.toString(), commonProjPrefix);
+                String newProject = StringUtils.removeStart(newWsFolder.toString(), commonProjPrefix);
+
+                throw new RuntimeException(String.format("Cannot automatically change uses of %s, since moving files between projects (from %s to %s) is not supported", currentLoc, currentProject, newProject));
+            }
+
+            PathConfig pcfg = getPathConfig.apply(currentWsFolder);
+            try {
+                IString currentName = VF.string(pcfg.getModuleName(currentLoc));
+                IString newName = VF.string(pcfg.getModuleName(newLoc));
+
+                writer.insert(VF.tuple(currentName, newName, addResources(pcfg)));
+            } catch (IOException e) {
+                throw new RuntimeException(e);
+            }
+        }
+
+        return writer.done();
+    }
+
+    private static String readFile(ISourceLocation loc, String charset) throws IOException {
```

Comment:
This is almost a copy from `ParametricTextDocumentService::getContents` with the exception that it doesn't look at the editor state, and doesn't use IOUtils but has the classic copy loop.

Is there a reason to keep this like this?

I would also advice against passing in the charset (unless we really know). In general, let the uri resolver take care of the charset.

Reply:
I will remove the explicit charset. Otherwise, this is more or less copied from `Prelude.readFile`, like you suggested. But I agree; the approach in `ParametricTextDocumentService::getContents` is a lot more readable. Will use that instead.